File: Python/50_Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            return 1/self.poww(x, abs(n))
        else:
            return self.poww(x, n)
        
    def poww(self, x, n):
        k = n//2
        if n == 0:
            return 1
        i = self.poww(x, abs(k))
        if n%2 == 0:
            if n < 0:
                return 1/(i*i)
            else:
                return  i * i
        else:
            if n < 0:
                return 1/(i*i*x)
            else:
                return  i * i * x

File: Java/50_Pow(x, n).java
class Solution {
    public double myPow(double x, int n) {
        double ans = Pow(x, Math.abs(n));
        if(n < 0) return 1.0/ans;
        else return ans;
    }
    
    public double Pow(double x, int n) {
        if(n == 0) return 1;
        double half = Pow(x, (int)n/2);
        if(n % 2 == 0) return half * half;
        else return x * half * half;
    }
    
}